### PR TITLE
Added pathfinding to the Application.

### PR DESCRIPTION
**Pull request by arekmano**
- Added the A* algorithm for pathfinding to the application.
- Added TiledMaps to the application, which will have to be tied with
  indoormaps somehow.

MCGA-100

### CODE REVIEW

File: app/src/main/java/com/concordia/mcga/utilities/pathfinding/PathFinder.java [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+        }
+        return returnList;
+    }
+
+    /**
+     * Runs an iteration of the A* algorithm.
+     * @param current - Currently examined tile.
+     * @throws Exception
+     */
+    private void nextIteration(PathFinderTile current) throws Exception {
```
Comment (adriannadiaz) [inline app/src/main/java/com/concordia/mcga/utilities/pathfinding/PathFinder.java]: Exception not needed.

File: app/src/main/java/com/concordia/mcga/utilities/pathfinding/PathFinder.java [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+    private PathFinderTile lowestOpen() throws Exception {
+        PathFinderTile lowest = PathFinderTile.MAX_COST;
+        for (PathFinderTile tile : openSet) {
+            if (tile.getCost() < lowest.getCost()) {
+                lowest = tile;
+            }
+        }
+        if (lowest.equals(PathFinderTile.MAX_COST)) {
+            throw new Exception("No Valid Path Found!");
```
Comment (adriannadiaz) [inline app/src/main/java/com/concordia/mcga/utilities/pathfinding/PathFinder.java]: Don't create an exception, return null instead and handle the null in the method calling lowestOpen()

File: app/src/main/java/com/concordia/mcga/utilities/pathfinding/PathFinder.java [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+                continue;
+            }
+            if (!closedSet.contains(tile)) {
+                if (openSet.contains(tile)) {
+                    int newDist = tile.calculateDistFromStart();
+                    if (newDist < tile.getDistFromStart()) {
+                        tile.setDistFromStart(newDist);
+                        tile.setParent(current);
+                        openSet.remove(tile);
+                        openSet.add(tile);
```
Comment (adriannadiaz) [inline app/src/main/java/com/concordia/mcga/utilities/pathfinding/PathFinder.java]: You add and remove the same tile from the same set? One of them should be openSet and the other one closedSet, no?

File: app/src/test/java/com/concordia/mcga/activities/PathFinderUnitTest.java [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+@RunWith(JUnit4.class)
+public class PathFinderUnitTest {
+    @Test
+    public void testShortestPath() throws Exception {
+        PathFinder finder = new PathFinder();
+
+        List<PathFinderTile> tiles = finder.shortestPath(1,1,4,4);
```
Comment (adriannadiaz) [inline app/src/test/java/com/concordia/mcga/activities/PathFinderUnitTest.java]: We should also test this code with at least two maps, one that has a solution, and the other that the destination is unreachable...

File: app/src/main/java/com/concordia/mcga/utilities/pathfinding/PathFinder.java (hunk 3 above)
Comment (arekmano) [inline app/src/main/java/com/concordia/mcga/utilities/pathfinding/PathFinder.java]: nope. a remnant of when I was trying to use a treeset.

File: app/src/test/java/com/concordia/mcga/activities/PathFinderUnitTest.java (hunk 4 above)
Comment (arekmano) [inline app/src/test/java/com/concordia/mcga/activities/PathFinderUnitTest.java]: yes. midnight push

File: app/src/main/java/com/concordia/mcga/utilities/pathfinding/PathFinder.java (hunk 1 above)
Comment (arekmano) [inline app/src/main/java/com/concordia/mcga/utilities/pathfinding/PathFinder.java]: eh sure

File: app/src/main/java/com/concordia/mcga/utilities/pathfinding/PathFinder.java (hunk 2 above)
Comment (arekmano) [inline app/src/main/java/com/concordia/mcga/utilities/pathfinding/PathFinder.java]: sure. though exception will be thrown somewhere

**Comment by Taimoorrana1** [comment]
Not sure how to test/review this without a map. The code itself looks good. Don't see any major error.

**Pull request closed without merging**